bootstrap: Include line numbers in debuginfo by default for library/compiler profile

jyn514 opened the pull request:
This only includes debuginfo for the relevant component, so for library and tool developers the compile time impact should be negligible. For compiler authors I think the improved backtraces are worth the slight compile time hit (roughly 10% slower).

- Set the default to LineTablesOnly. This isn't enough debuginfo to run perf, but it's enough to
get full backtraces.
- Don't build debuginfo for build scripts and proc macros. This lowers the total disk space for stage0-rustc from 5.4 to 5.3 GB, or by 120 MB.
- Fix a bug in deserialization where string values would never be deserialized

  For reasons I don't quite understand, using `&str` unconditionally failed:
    ```
    thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { inner: Error { inner: TomlError { message: "data did not match any variant of untagged enum StringOrInt", original: None, keys: [], span: None } } }', src/main.rs:15:49
    ```

Mark-Simulacrum commented:
My recollection is that this increases the disk size pretty considerably as well, could you measure that while you get the timing estimate?

jyn514 commented:
before:

```
; x build --stage 0 compiler
Building stage0 compiler artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 8m 13s

; ls build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/{librustc_driver.so,rustc-main} -lh
-rwxrwxr-x 4 jnelson jnelson 115M Nov 28 13:20 build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/librustc_driver.so
-rwxrwxr-x 3 jnelson jnelson  17K Nov 28 13:20 build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/rustc-main
```

after:

```
; x build --stage 0 compiler
Building stage0 compiler artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 10m 55s

; ls build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/{librustc_driver.so,rustc-main} -lh
-rwxrwxr-x 4 jnelson jnelson 529M Nov 28 13:36 build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/librustc_driver.so
-rwxrwxr-x 3 jnelson jnelson  22K Nov 28 13:36 build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/rustc-main
```

jyn514 commented:
I wonder if it makes sense to only do this for the `tools` and `library` profiles?

saethlin commented:
I think I have a system with a bit more parallelism. I see a much larger % slowdown, and turning on `use-lld = true` removes most of it.

I do not think this degree of slowdown is acceptable for a default, especially because evidence suggests it is mostly loaded into running the linker, which will always be done. If we could have `use-lld = true` on by default I think the regression _might_ be tolerable.

> I wonder if it makes sense to only do this for the `tools` and `library` profiles?

I think @thomcc previously commented that the benefit for `library` is dubious. I defer to more experienced library contributors there. I also think I spend a lot more time profiling Miri than anyone else, and the iteration speed on it is already rather brutal, so similarly I'm cautious about penalizing the workflows of others for my own convenience.

jyn514 commented:
I thought @thomcc was in favor of this change for library.

Mark-Simulacrum commented:
It's really unfortunate to increase the library by ~5x, to half a gigabyte(!), even aside from the potential wall-time slowdown.

I don't think we should do this by default at this time, it's too expensive a default. I do wonder whether the debuginfo really *should* be taking that much room -- naively, one might expect that a "line tables only" version should be able to be roughly on-par with what the source code (actually better!) itself takes up, whereas this is much bigger. Maybe we're including not just line table information but something more (e.g., types of some kind)?

cc @rust-lang/wg-debugging -- I wonder if folks here have thoughts on what we can do to give a nicer experience at lower cost, or whether we think debuginfo is the wrong way to go, etc.

michaelwoerister commented:
The 5x size increase does seem fishy.

bors commented:
:umbrella: The latest upstream changes (presumably #107241) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 committed:
Include line numbers in debuginfo by default for developers

- Set the default to LineTablesOnly. This isn't enough debuginfo to run perf, but it's enough to
get full backtraces.
- Don't build debuginfo for build scripts and proc macros. This lowers the total disk space for stage0-rustc from 5.4 to 5.3 GB, or by 120 MB.
- Fix a bug in deserialization where string values would never be deserialized

  For reasons I don't quite understand, using `&str` unconditionally failed:
    ```
    thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { inner: Error { inner: TomlError { message: "data did not match any variant of untagged enum StringOrInt", original: None, keys: [], span: None } } }', src/main.rs:15:49
    ```

jyn514 commented:
It turns out the vast majority of the size (over 350 GB!) comes from the `.debug_info`, `.debug_pubnames`, and `.debug_str` sections, not the line info itself: https://rust-lang.zulipchat.com/#narrow/stream/317568-t-compiler.2Fwg-debugging/topic/investigating.20debuginfo.20size
Apparently there's tooling depending on those sections to exist :facepalm: so we can't remove it: https://github.com/rust-lang/rust/issues/64405

jyn514 commented:
https://github.com/rust-lang/rust/issues/69074 also looks related.

jyn514 commented:
Ok, I added a new unstable flag that stops generating extra debuginfo when only line-level info is set, and enabled that and compression when building rustc itself. After that change, `librustc_driver.so` is 154 MB on my computer, which seems roughly comparable to the 131 MB used by the version without debuginfo. Do you think that's small enough we can land the change in defaults?

jyn514 commented:
ah, I just realized the reduced size won't take effect for most contributors until the new flag hits beta - I can split it into a separate PR so it lands without having to change the defaults at the same time.

jyn514 commented:
Wall time before:
```
Building compiler artifacts (stage1 -> stage2)
    Finished release [optimized] target(s) in 1m 58s
Assembling stage2 compiler
Build completed successfully in 0:01:59
```
and after:
```
Building compiler artifacts (stage1 -> stage2)
    Finished release [optimized + debuginfo] target(s) in 2m 13s
Assembling stage2 compiler
Build completed successfully in 0:02:13
```

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4456146349/jobs/7826396449) [(plain)](https://github.com/rust-lang/rust/commit/f55c1db45321e554be4c3183757f8a35239bf80c/checks/12101581828/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:24cb9080177205b6e8c946b17badbe402adc938f)
Download action repository 'rust-lang/simpleinfra@master' (SHA:3fb2b44a4eaebb9ed8086446bde46c27199ef5ed)
Complete job name: PR (x86_64-gnu-llvm-14, false, ubuntu-20.04-xl)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-llvm-14
---

---- [ui] tests/rustdoc-ui/z-help.rs stdout ----
diff of stdout:

45     -Z                 extra-const-ub-checks=val -- turns on more checks to detect const UB, which can be slow (default: no)
46     -Z                           fewer-names=val -- reduce memory use by retaining fewer names within compilation artifacts (LLVM-IR) (default: no)
47     -Z                   flatten-format-args=val -- flatten nested format_args!() and literals into a simplified format_args!() call (default: no)
+     -Z                  force-full-debuginfo=val -- force all codegen-units to have full debuginfo even for -C debuginfo=1. see #64405 (default: yes)
48     -Z            force-unstable-if-unmarked=val -- force all crates to be `rustc_private` unstable (default: no)
49     -Z                                  fuel=val -- set the optimization fuel quota for a crate
50     -Z                     function-sections=val -- whether each function should go in its own section

The actual stdout differed from the expected stdout.
The actual stdout differed from the expected stdout.
Actual stdout saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/z-help/z-help.stdout
To only update this specific test, also pass `--test-args z-help.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "/checkout/tests/rustdoc-ui/z-help.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/z-help" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/z-help/auxiliary" "-Zhelp"
--- stdout -------------------------------
-Z                        allow-features=val -- only allow the listed language features to be enabled in code (comma separated)
    -Z                     always-encode-mir=val -- encode MIR of all functions into the crate metadata (default: no)
    -Z                          asm-comments=val -- generate comments into the assembly (may change behavior) (default: no)
    -Z                     assert-incr-state=val -- assert that the incremental cache is in given state: either `loaded` or `not-loaded`.
    -Z             assume-incomplete-release=val -- make cfg(version) treat the current version as incomplete (default: no)
    -Z                    binary-dep-depinfo=val -- include artifacts (sysroot, crate dependencies) used during compilation in dep-info (default: no)
    -Z                           box-noalias=val -- emit noalias metadata for box (default: yes)
    -Z                     branch-protection=val -- set options for branch target identification and pointer authentication on AArch64
    -Z                         cf-protection=val -- instrument control-flow architecture protection
    -Z             cgu-partitioning-strategy=val -- the codegen unit partitioning strategy to use
    -Z                       codegen-backend=val -- the backend to use
    -Z                           combine-cgu=val -- combine CGUs into a single one
    -Z                            crate-attr=val -- inject the given attribute in the crate
    -Z              debug-info-for-profiling=val -- emit discriminators and other data necessary for AutoFDO
    -Z                          debug-macros=val -- emit line numbers debug info inside macros (default: no)
    -Z               deduplicate-diagnostics=val -- deduplicate identical diagnostics (default: yes)
    -Z                dep-info-omit-d-target=val -- in dep-info output, omit targets for tracking dependencies of the dep-info files themselves (default: no)
    -Z                             dep-tasks=val -- print tasks that execute and the color their dep node gets (requires debug build) (default: no)
    -Z                      diagnostic-width=val -- set the current output width for diagnostic truncation
    -Z                               dlltool=val -- import library generation tool (windows-gnu only)
    -Z               dont-buffer-diagnostics=val -- emit diagnostics rather than buffering (breaks NLL error downgrading, sorting) (default: no)
    -Z                         drop-tracking=val -- enables drop tracking in generators (default: no)
    -Z                     drop-tracking-mir=val -- enables drop tracking on MIR in generators (default: no)
    -Z                      dual-proc-macros=val -- load proc macros for both target and host, but only link to the target (default: no)
    -Z                        dump-dep-graph=val -- dump the dependency graph to $RUST_DEP_GRAPH (default: /tmp/dep_graph.gv) (default: no)
    -Z                dump-drop-tracking-cfg=val -- dump drop-tracking control-flow graph as a `.dot` file (default: no)
    -Z                              dump-mir=val -- dump MIR state to file.
        `val` is used to select which passes and functions to dump. For example:
        `all` matches all passes and functions,
        `foo` matches all passes for functions whose name contains 'foo',
        `foo & ConstProp` only the 'ConstProp' pass for function names containing 'foo',
        `foo | bar` all passes for function names containing 'foo' or 'bar'.
    -Z                     dump-mir-dataflow=val -- in addition to `.mir` files, create graphviz `.dot` files with dataflow results (default: no)
    -Z                          dump-mir-dir=val -- the directory the MIR is dumped into (default: `mir_dump`)
    -Z          dump-mir-exclude-pass-number=val -- exclude the pass number when dumping MIR (used in tests) (default: no)
    -Z                     dump-mir-graphviz=val -- in addition to `.mir` files, create graphviz `.dot` files (and with `-Z instrument-coverage`, also create a `.dot` file for the MIR-derived coverage graph) (default: no)
    -Z                     dump-mir-spanview=val -- in addition to `.mir` files, create `.html` files to view spans for all `statement`s (including terminators), only `terminator` spans, or computed `block` spans (one span encompassing a block's terminator and all statements). If `-Z instrument-coverage` is also enabled, create an additional `.html` file showing the computed coverage spans.
    -Z                       dump-mono-stats=val -- output statistics about monomorphization collection
    -Z                dump-mono-stats-format=val -- the format to use for -Z dump-mono-stats (`markdown` (default) or `json`)
    -Z                         dwarf-version=val -- version of DWARF debug information to emit (default: 2 or 4, depending on platform)
    -Z                             dylib-lto=val -- enables LTO for dylib crate type
    -Z                      emit-stack-sizes=val -- emit a section containing stack size metadata (default: no)
    -Z                         emit-thin-lto=val -- emit the bc module with thin LTO info (default: yes)
    -Z             export-executable-symbols=val -- export symbols from executables, as if they were dynamic libraries
    -Z                 extra-const-ub-checks=val -- turns on more checks to detect const UB, which can be slow (default: no)
    -Z                           fewer-names=val -- reduce memory use by retaining fewer names within compilation artifacts (LLVM-IR) (default: no)
    -Z                   flatten-format-args=val -- flatten nested format_args!() and literals into a simplified format_args!() call (default: no)
    -Z                  force-full-debuginfo=val -- force all codegen-units to have full debuginfo even for -C debuginfo=1. see #64405 (default: yes)
    -Z            force-unstable-if-unmarked=val -- force all crates to be `rustc_private` unstable (default: no)
    -Z                                  fuel=val -- set the optimization fuel quota for a crate
    -Z                     function-sections=val -- whether each function should go in its own section
    -Z                  future-incompat-test=val -- forces all lints to be future incompatible, used for internal testing (default: no)
    -Z                                gcc-ld=val -- implementation of ld used by cc
    -Z                    graphviz-dark-mode=val -- use dark-themed colors in graphviz output (default: no)
    -Z                         graphviz-font=val -- use the given `fontname` in graphviz output; can be overridden by setting environment variable `RUSTC_GRAPHVIZ_FONT` (default: `Courier, monospace`)
    -Z                             hir-stats=val -- print some statistics about AST and HIR (default: no)
    -Z              human-readable-cgu-names=val -- generate human-readable, predictable names for codegen units (default: no)
    -Z                      identify-regions=val -- display unnamed regions as `'<id>`, using a non-ident unique id (default: no)
    -Z              incremental-ignore-spans=val -- ignore spans during ICH computation -- used for testing (default: no)
    -Z                      incremental-info=val -- print high-level information about incremental reuse (or the lack thereof) (default: no)
    -Z            incremental-relative-spans=val -- hash spans relative to their parent item for incr. comp. (default: no)
    -Z                incremental-verify-ich=val -- verify incr. comp. hashes of green query instances (default: no)
    -Z                    inline-in-all-cgus=val -- control whether `#[inline]` functions are in all CGUs
    -Z                           inline-llvm=val -- enable LLVM inlining (default: yes)
    -Z                            inline-mir=val -- enable MIR inlining (default: no)
    -Z             inline-mir-hint-threshold=val -- inlining threshold for functions with inline hint (default: 100)
    -Z                  inline-mir-threshold=val -- a default MIR inlining threshold (default: 50)
    -Z                           input-stats=val -- gather statistics about the input (default: no)
    -Z                   instrument-coverage=val -- instrument the generated code to support LLVM source-based code coverage reports (note, the compiler build config must include `profiler = true`); implies `-C symbol-mangling-version=v0`. Optional values are:
        `=all` (implicit value)
        `=except-unused-generics`
        `=except-unused-functions`
        `=off` (default)
    -Z                     instrument-mcount=val -- insert function instrument code for mcount-based tracing (default: no)
    -Z                       instrument-xray=val -- insert function instrument code for XRay-based tracing (default: no)
         Optional extra settings:
         `=always`
         `=never`
         `=ignore-loops`
         `=instruction-threshold=N`
         `=skip-entry`
         `=skip-exit`
         Multiple options can be combined with commas.
    -Z                     keep-hygiene-data=val -- keep hygiene data after analysis (default: no)
    -Z                           layout-seed=val -- seed layout randomization
    -Z                       link-directives=val -- honor #[link] directives in the compiled crate (default: yes)
    -Z                 link-native-libraries=val -- link native libraries in the linker invocation (default: yes)
    -Z                             link-only=val -- link the `.rlink` file generated by `-Z no-link` (default: no)
    -Z                          llvm-plugins=val -- a list LLVM plugins to enable (space separated)
    -Z                       llvm-time-trace=val -- generate JSON tracing data file from LLVM data (default: no)
    -Z                       location-detail=val -- what location details should be tracked when using caller_location, either `none`, or a comma separated list of location details, for which valid options are `file`, `line`, and `column` (default: `file,line,column`)
    -Z lower-impl-trait-in-trait-to-assoc-ty=val -- modify the lowering strategy for `impl Trait` in traits so that they are lowered to generic associated types
    -Z                                    ls=val -- list the symbols defined by a library crate (default: no)
    -Z                       macro-backtrace=val -- show macro backtraces (default: no)
    -Z           maximal-hir-to-mir-coverage=val -- save as much information as possible about the correspondence between MIR and HIR as source scopes (default: no)
    -Z                       merge-functions=val -- control the operation of the MergeFunctions LLVM pass, taking the same values as the target option of the same name
    -Z                            meta-stats=val -- gather metadata statistics (default: no)
    -Z                        mir-emit-retag=val -- emit Retagging MIR statements, interpreted e.g., by miri; implies -Zmir-opt-level=0 (default: no)
    -Z                     mir-enable-passes=val -- use like `-Zmir-enable-passes=+DestProp,-InstCombine`. Forces the specified passes to be enabled, overriding all other checks. Passes that are not specified are enabled or disabled by other flags as usual.
    -Z                         mir-opt-level=val -- MIR optimization level (0-4; default: 1 in non optimized builds and 2 in optimized builds)
    -Z      mir-pretty-relative-line-numbers=val -- use line numbers relative to the function in mir pretty printing
    -Z                       move-size-limit=val -- the size at which the `large_assignments` lint starts to be emitted
    -Z                       mutable-noalias=val -- emit noalias metadata for mutable references (default: yes)
    -Z                             nll-facts=val -- dump facts from NLL analysis into side files (default: no)
    -Z                         nll-facts-dir=val -- the directory the NLL facts are dumped into (default: `nll-facts`)
    -Z                           no-analysis=val -- parse and expand the source, but run no analysis
    -Z                            no-codegen=val -- run all passes except codegen; no output
    -Z            no-generate-arange-section=val -- omit DWARF address ranges that give faster lookups
    -Z                        no-jump-tables=val -- disable the jump tables and lookup tables that can be generated from a switch case lowering
    -Z                         no-leak-check=val -- disable the 'leak check' for subtyping; unsound, but useful for tests
    -Z                               no-link=val -- compile without linking
    -Z                      no-parallel-llvm=val -- run LLVM in non-parallel mode (while keeping codegen-units and ThinLTO)
    -Z                   no-profiler-runtime=val -- prevent automatic injection of the profiler_builtins crate
    -Z               no-unique-section-names=val -- do not use unique names for text and data sections when -Z function-sections is used
    -Z                        normalize-docs=val -- normalize associated items in rustdoc when generating documentation
    -Z                                   oom=val -- panic strategy for out-of-memory handling
    -Z                osx-rpath-install-name=val -- pass `-install_name @rpath/...` to the macOS linker (default: no)
    -Z                   packed-bundled-libs=val -- change rlib format to store native libraries as archives
    -Z                     panic-abort-tests=val -- support compiling tests with panic=abort (default: no)
    -Z                         panic-in-drop=val -- panic strategy for panics in drops
    -Z                            parse-only=val -- parse only; do not compile, assemble, or link (default: no)
    -Z                            perf-stats=val -- print some performance-related statistics (default: no)
    -Z                                   plt=val -- whether to use the PLT when calling into shared libraries;
        only has effect for PIC code on systems with ELF binaries
        (default: PLT is disabled if full relro is enabled)
    -Z                              polonius=val -- enable polonius-based borrow-checker (default: no)
    -Z                          polymorphize=val -- perform polymorphization analysis
    -Z                          pre-link-arg=val -- a single extra argument to prepend the linker invocation (can be used several times)
    -Z                         pre-link-args=val -- extra arguments to prepend to the linker invocation (space separated)
    -Z         precise-enum-drop-elaboration=val -- use a more precise version of drop elaboration for matches on enums (default: yes). This results in better codegen, but has caused miscompilations on some tier 2 platforms. See #77382 and #74551.
    -Z                            print-fuel=val -- make rustc print the total optimization fuel used by a crate
    -Z                     print-llvm-passes=val -- print the LLVM optimization passes being run (default: no)
    -Z                      print-mono-items=val -- print the result of the monomorphization collection pass
    -Z                      print-type-sizes=val -- print layout information for each type encountered (default: no)
    -Z                  proc-macro-backtrace=val -- show backtraces for panics during proc-macro execution (default: no)
    -Z         proc-macro-execution-strategy=val -- how to run proc-macro code (default: same-thread)
    -Z                               profile=val -- insert profiling code (default: no)
    -Z                      profile-closures=val -- profile size of closures
    -Z                          profile-emit=val -- file path to emit profiling data at runtime when using 'profile' (default based on relative source path)
    -Z                    profile-sample-use=val -- use the given `.prof` file for sampled profile-guided optimization (also known as AutoFDO)
    -Z                      profiler-runtime=val -- name of the profiler runtime crate to automatically inject (default: `profiler_builtins`)
    -Z                       query-dep-graph=val -- enable queries of the dependency graph for regression testing (default: no)
    -Z                      randomize-layout=val -- randomize the layout of types (default: no)
    -Z                 relax-elf-relocations=val -- whether ELF relocations can be relaxed
    -Z                           relro-level=val -- choose which RELRO level to use
    -Z                      remap-cwd-prefix=val -- remap paths under the current working directory to this path prefix
    -Z                   report-delayed-bugs=val -- immediately print bugs registered with `delay_span_bug` (default: no)
    -Z                             sanitizer=val -- use a sanitizer
    -Z        sanitizer-memory-track-origins=val -- enable origins tracking in MemorySanitizer
    -Z                     sanitizer-recover=val -- enable recovery for selected sanitizers
    -Z                saturating-float-casts=val -- make float->int casts UB-free: numbers outside the integer type's range are clipped to the max/min integer respectively, and NaN is mapped to 0 (default: yes)
    -Z                          self-profile=val -- run the self profiler and output the raw event data
    -Z                  self-profile-counter=val -- counter used by the self profiler (default: `wall-time`), one of:
        `wall-time` (monotonic clock, i.e. `std::time::Instant`)
        `instructions:u` (retired instructions, userspace-only)
        `instructions-minus-irqs:u` (subtracting hardware interrupt counts for extra accuracy)
    -Z                   self-profile-events=val -- specify the events recorded by the self profiler;
        for example: `-Z self-profile-events=default,query-keys`
        all options: none, all, default, generic-activity, query-provider, query-cache-hit
                     query-blocked, incr-cache-load, incr-result-hashing, query-keys, function-args, args, llvm, artifact-sizes
    -Z                        share-generics=val -- make the current crate share its generic instantiations
    -Z                             show-span=val -- show spans for compiler debugging (expr|pat|ty)
    -Z       simulate-remapped-rust-src-base=val -- simulate the effect of remap-debuginfo = true at bootstrapping by remapping path to rust's source base directory. only meant for testing purposes
    -Z                            span-debug=val -- forward proc_macro::Span's `Debug` impl to `Span`
    -Z                     span-free-formats=val -- exclude spans when debug-printing compiler state (default: no)
    -Z                  split-dwarf-inlining=val -- provide minimal debug info in the object/executable to facilitate online symbolication/stack traces in the absence of .dwo/.dwp files when using Split DWARF
    -Z                      split-dwarf-kind=val -- split dwarf variant (only if -Csplit-debuginfo is enabled and on relevant platform)
        (default: `split`)

        `split`: sections which do not require relocation are written into a DWARF object (`.dwo`)
                 file which is ignored by the linker
        `single`: sections which do not require relocation are written into object file but ignored
                  by the linker
    -Z                    src-hash-algorithm=val -- hash algorithm of source files in debug info (`md5`, `sha1`, or `sha256`)
    -Z                       stack-protector=val -- control stack smash protection strategy (`rustc --print stack-protector-strategies` for details)
    -Z                    strict-init-checks=val -- control if mem::uninitialized and mem::zeroed panic on more UB
    -Z                                 strip=val -- tell the linker which information to strip (`none` (default), `debuginfo` or `symbols`)
    -Z               symbol-mangling-version=val -- which mangling version to use for symbol names ('legacy' (default) or 'v0')
    -Z                                 teach=val -- show extended diagnostic help (default: no)
    -Z                             temps-dir=val -- the directory the intermediate files are written to
    -Z                         terminal-urls=val -- use the OSC 8 hyperlink terminal specification to print hyperlinks in the compiler output
    -Z                               thinlto=val -- enable ThinLTO when possible
    -Z                         thir-unsafeck=val -- use the THIR unsafety checker (default: no)
    -Z                               threads=val -- use a thread pool with N threads
    -Z                      time-llvm-passes=val -- measure time of each LLVM pass (default: no)
    -Z                           time-passes=val -- measure time of each rustc pass (default: no)
    -Z                 tiny-const-eval-limit=val -- sets a tiny, non-configurable limit for const eval; useful for compiler tests
    -Z                             tls-model=val -- choose the TLS model to use (`rustc --print tls-models` for details)
    -Z                          trace-macros=val -- for every macro invocation, print its name and arguments (default: no)
    -Z                     track-diagnostics=val -- tracks where in rustc a diagnostic was emitted
    -Z                          trait-solver=val -- specify the trait solver mode used by rustc (default: classic)
    -Z              translate-additional-ftl=val -- additional fluent translation to preferentially use (for testing translation)
    -Z      translate-directionality-markers=val -- emit directionality isolation markers in translated diagnostics
    -Z                        translate-lang=val -- language identifier for diagnostic output
    -Z translate-remapped-path-to-local-path=val -- translate remapped paths into local paths when possible (default: yes)
    -Z                      trap-unreachable=val -- generate trap instructions for unreachable intrinsics (default: use target setting, usually yes)
    -Z                      treat-err-as-bug=val -- treat error number `val` that occurs as bug
    -Z                 trim-diagnostic-paths=val -- in diagnostics, use heuristics to shorten paths referring to items
    -Z                              tune-cpu=val -- select processor to schedule for (`rustc --print target-cpus` for details)
    -Z                            ui-testing=val -- emit compiler diagnostics in a form suitable for UI testing (default: no)
    -Z          uninit-const-chunk-threshold=val -- allow generating const initializers with mixed init/uninit chunks, and set the maximum number of chunks for which this is allowed (default: 16)
    -Z        unleash-the-miri-inside-of-you=val -- take the brakes off const evaluation. NOTE: this is unsound (default: no)
    -Z                              unpretty=val -- present the input source, unstable (and less-pretty) variants;
        `normal`, `identified`,
        `expanded`, `expanded,identified`,
        `expanded,hygiene` (with internal representations),
        `ast-tree` (raw AST before expansion),
        `ast-tree,expanded` (raw AST after expansion),
        `hir` (the HIR), `hir,identified`,
        `hir,typed` (HIR with types for each node),
        `hir-tree` (dump the raw HIR),
        `mir` (the MIR), or `mir-cfg` (graphviz formatted MIR)
    -Z                      unsound-mir-opts=val -- enable unsound and buggy MIR optimizations (default: no)
    -Z                      unstable-options=val -- adds unstable command line options to rustc interface (default: no)
    -Z                     use-ctors-section=val -- use legacy .ctors section for initializers rather than .init_array
    -Z                          validate-mir=val -- validate MIR after each transformation
    -Z                               verbose=val -- in general, enable more debug printouts (default: no)
    -Z                        verify-llvm-ir=val -- verify LLVM IR (default: no)
    -Z          virtual-function-elimination=val -- enables dead virtual function elimination optimization. Requires `-Clto[=[fat,yes]]`
    -Z                       wasi-exec-model=val -- whether to build a wasi command or reactor
stderr: none



```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #109808) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 commented:
waiting on me to test https://github.com/rust-lang/rust/pull/110221#issuecomment-1575081397

jyn514 commented:
> I think a sense of what the impact is on the full build directory with and without debuginfo would help make a decision on whether this makes sense as a default (the PR contains numbers for librustc_driver only).

Without debuginfo:
```
; CARGOFLAGS=--timings x b --stage 0 compiler
Building compiler artifacts (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 1m 07s
; usage build/host/stage0-rustc
3.4G    build/host/stage0-rustc
3.0G    build/host/stage0-rustc/x86_64-unknown-linux-gnu
355M    build/host/stage0-rustc/release
244K    build/host/stage0-rustc/cargo-timings
```

With debuginfo:
```
; CARGOFLAGS=--timings x b --stage 0 compiler
Building compiler artifacts (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 1m 13s
; usage build/host/stage0-rustc
4.4G    build/host/stage0-rustc
4.0G    build/host/stage0-rustc/x86_64-unknown-linux-gnu
401M    build/host/stage0-rustc/release
256K    build/host/stage0-rustc/cargo-timings
```

For comparison, `build/host/stage0-std` takes up 500 MB total in both cases.

Both of these were taken with the default settings from `profile = compiler` on b2b34bd83192c3d16c88655158f7d8d612513e88.

jyn514 commented:
oh, i just realized we can cut down most of the added space in `release` by disabling debuginfo for build dependencies - that should save 50 mb or so.

Mark-Simulacrum commented:
r=me, unless you want to do more here.

@rustbot author

jyn514 commented:
This had a semantic conflict with #112528 - I've rebased and pushed a commit that fixes the conflict as well as not building debuginfo for build scripts.

jyn514 commented:
err it's not actually correct, it changed the default for `debug = true` instead of the per-profile default

jyn514 commented:
something here is broken and i don't have time to investigate it right now
```
; ls -lh build/host/stage0-rustc/x86_64-unknown-linux-gnu/release/librustc_driver.so 
-rwxrwxr-x 4 jyn 278M Jun 23 22:35 build/host/stage0-rustc/x86_64-unknown-linux-gnu/release/librustc_driver.so
```
this was *significantly* smaller last time i measure i tand i don't know what changed.

bors commented:
:umbrella: The latest upstream changes (presumably #112756) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
